### ISSUE TITLE
Add record button to AnimationPlayer

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#169]
**Describe the project you are working on:**
Projects that require me to create many animations, tracks, keys.

**Describe the problem or limitation you are having in your project:**
Creating animations, tracks and key is a very tedious process. Especially so when trying to animate not just one property of an object, but a few.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
If we had a record feature for the AnimationPlayer, I could hit "record", make a change to any object, hit pause or stop when I'm done, and only fine tune the timing and values if needed. 

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![record_button](https://user-images.githubusercontent.com/47016402/67154161-62cda900-f2f7-11e9-8b05-0b45c73b5efb.gif)

**Describe implementation detail for your proposal (in code), if possible:**
I only got a C-- in the "How to turn your computer on and off" class. So I'm not quite on the level of C++ yet. 

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I suppose it would be used a lot, as it would make animating anything a whole lot faster and easier.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It's another UX improvement to the AnimationPlayer. I don't think you should need go to the asset library to make existing tools efficient to use.

**Comment by Calinou** [comment]
How does this differ from the autokeying option (the "car" icon at the top of the 2D view when AnimationPlayer is selected)?

![image](https://user-images.githubusercontent.com/180032/67156587-cd470f00-f320-11e9-9811-72127647af04.png)

Note that this feature *is* present in 3.1.1, but the icon wasn't added in time, which means you have to click the invisible space where the icon would be.

**Comment by KoBeWi** [comment]
> How does this differ from the autokeying option

I think it's about recording a set of values over time, i.e. you change property and it's added at current time, but time advances automatically (not sure if this is a case with current auto-keying)

**Comment by golddotasksquestions** [comment]
@Calinou 
Neither 3.1.1 not 3.2 alpha2 Win64 has a visible or invisible car symbol up there, so I really don't know how to answer this question.

![no_car](https://user-images.githubusercontent.com/47016402/67159398-69364200-f344-11e9-90fa-85742cbd6968.gif)

Even if such a feature already excited (which I cannot test, as it is not there), having it as (in)visible car up there is really of no good. Please for the love of everything sacred, lets not question incredibly basic well known user behavior. A red disc next to a play and stop is what everyone knows and immediately recognizes as a record button. Please, oh please, let's not make Godot usability worse by inventing our own little way of doing things when there already are well known world wide spread user behaviors existing.

**Comment by golddotasksquestions** [comment]
Found the car in 3.2. alpha2, not sure what I did to not make it appear before. 
Anyway, so I can answer your question @Calinou:

1. It's a car. A recording button asks for a recording sympol (red disk next to play and stop)
2. Even by design there are preexisting condition to meet before it does anything (create or use and existing animation, create or use existing track, create first key manually.)
3. Even if those conditions are met, it does not seem to record anything at all. Only if there is another key set, it does seem to do something. However it does not record any new keys (which is what this issue is about), only the last key value I already set seems to change:

![car_not_recording_32](https://user-images.githubusercontent.com/47016402/67162186-03f14980-f362-11e9-96bf-dac457510bd4.gif)

4. Thinking it may only override existing key values (which would still not be what this issue is about but still having better than no recording at all), I tried adding keys, pressing the car and play hoping to override existing keys (it's the same without play):
![car_not_recording2_32](https://user-images.githubusercontent.com/47016402/67162343-597a2600-f363-11e9-8d53-e0994fb033ce.gif)

So to summarize: That car does not look like a recording feature, it's not at the place where I would expect a recording feature, it does not perform at all like a recording feature, so in total it does nothing I would expect from a feature outlined in this proposal.

**Comment by avencherus** [comment]
> Found the car in 3.2. alpha2, not sure what I did to not make it appear before.

It was directly committed https://github.com/godotengine/godot/commit/296c74072c0f4d71567c425a94d9502a68b6e177 with the icon missing initially, then added https://github.com/godotengine/godot/commit/5d65100669e8a8ece09677feb744c69b4466aa90

Only the first commit was taken into 3.1,1 which went to release shortly after.  So nothing you did or didn't do.  It should be appearing in the next 3.1 release if the PR above is merged before then.

Also, someone once suggested this.  Not sure if it went anywhere.
![image](https://user-images.githubusercontent.com/13004169/67163861-d49c0600-f37c-11e9-9edf-b13b510c7e55.png)

**Comment by vyshliy** [comment]
I want to propose moving the record button directly to the animation player, exactly as originally proposed. Everything highlighted in green in the screenshot refers directly to the selected node, except for the record button. The record button is related to the state of the animation player itself. In addition, this is a much more familiar location for users of any other software. In the context of this thread #4273, everything in the green area will probably be moved to the inspector, making the state of the record button even less clear, while the position in the player is clear and unambiguous
![Screenshot_20230412_152305](https://user-images.githubusercontent.com/72736632/231495409-52e4b395-21ab-4da6-a717-bb30e609db1e.png)

**Comment by igtCWrightson** [comment]
Is there a reason this functionality does not exist for 3D objects? And I agree with the original post, the record button should be present in the Animation tab as shown in the first screenshot.

**Comment by Calinou** [comment]
> Is there a reason this functionality does not exist for 3D objects?

I think the only reason is a technical/code organization issue. It was never implemented in the 3D editor, as it requires duplicating the code that was added in the 2D editor.

**Comment by igtCWrightson** [comment]
> I think the only reason is a technical/code organization issue. It was never implemented in the 3D editor, as it requires duplicating the code that was added in the 2D editor.

Good to know. Thank you. I would LOVE to have this feature in the 3D mode. Maybe it's not too hard to bring over from 2D.

**Comment by tobiasBora** [comment]
I’d also be fairly interested (this is possible possible in blender), for both 2D & 3D, and ideally any property (or maybe we can bound them to bones and then just animate the position of the bones?). I tried the record button, but even in 3D it is not doing anything while playing:

![Peek 30-05-2024 11-44](https://github.com/godotengine/godot-proposals/assets/2164118/b968eac8-1605-4f01-bb5c-c5c3c3e7250f)

**Comment by Creta5164** [comment]
I'll leave a detour into my method of adding keyframes programatically in tool for those who may find it later via search, as a public service, for those who want tooling that ties into the `AnimationPlayer` editor:

What I wanted to do with this idea was to streamline my composition workflow, which is exactly what this proposal does.

https://github.com/user-attachments/assets/6dd09727-9199-40af-9c2b-0ba081a52a06

```gdscript
var inspector     := EditorInterface.get_inspector()
var edited_object := inspector.get_edited_object()

EditorInterface.inspect_object(target_node)
inspector.property_keyed.emit.call_deferred(property, value, false)

EditorInterface.inspect_object.call_deferred(edited_object)
```

1. Get the `EditorInspector` from the `EditorInterface`, and cache the object currently being edited.
2. Specify the node you want to add keyframes to the track in the Animation Player that you are currently editing via `EditorInterface.inspect_object`.
3. Emit the `property_keyed` signal through the `EditorInspector`.
4. Revert back to the object you cached in 1 by pointing it back to `EditorInterface.inspect_object`.

Obviously, this isn't a clean way to do it (for example, if you have an undo action, you'll have to do it at least twice to undo it), but I could accomplish the needs addressed in this proposal as an external tool script.

Since this is also related to the editor tooling, I'll put #6828 in the content to make it easier to track the reference, hopefully more people will pay attention to this proposal and linked proposal.
